src/SourceLocator/Ast/Parser/MemoizingParser.php

```diff
@@ -46,11 +43,11 @@ public function parse(string $code, ?ErrorHandler $errorHandler = null): ?array
         $hash = hash('sha256', $code) . ':' . strlen($code);
 
         if (array_key_exists($hash, $this->sourceHashToAst)) {
-            return $this->jsonDecoder->decode($this->sourceHashToAst[$hash]);
+            return $this->sourceHashToAst[$hash];
         }
 
         $ast                          = $this->wrappedParser->parse($code, $errorHandler);
-        $this->sourceHashToAst[$hash] = json_encode($ast, JSON_PARTIAL_OUTPUT_ON_ERROR | JSON_PRESERVE_ZERO_FRACTION);
+        $this->sourceHashToAst[$hash] = $ast;
```

Comment:
The reason why this is done in such way is that the AST **must** be made immutable.

`serialize()` could also work: just need to see the performance impact - @ondrejmirtes and @kukulich did some work on this before.